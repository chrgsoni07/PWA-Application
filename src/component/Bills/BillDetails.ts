export type BillDetails = {
  newTotal: number;
  oldTotal: number;
  oldNewDifference: number;
  amountPayable: number;
  discount: number;
  paid: number;
  due: number;
};
