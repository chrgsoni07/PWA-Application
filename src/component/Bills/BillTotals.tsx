import { InputNumber } from "primereact/inputnumber";
import React from "react";
import { defaultNum } from "utils/number.utils";
export function BillTotals({
  onDiscoutChange,
  onAmountPaidChange,
  billDetails,
}: any) {
  return (
    <div className="p-fluid p-formgrid p-grid">
      <div className="p-field p-col">
        <label htmlFor="totalNew"> Total new </label>
        <InputNumber id="totalNew" value={defaultNum(billDetails.newTotal)} />
      </div>

      <div className="p-field p-col">
        <label htmlFor="totalOld">Total old </label>
        <InputNumber id="totalOld" value={defaultNum(billDetails.oldTotal)} />
      </div>

      <div className="p-field p-col">
        <label htmlFor="oldNewDifference"> Old New Difference </label>
        <InputNumber
          id="oldNewDifference"
          value={defaultNum(billDetails.oldNewDifference)}
        />
      </div>

      <div className="p-field p-col">
        <label htmlFor="discount">Discount</label>
        <InputNumber
          id="discount"
          value={defaultNum(billDetails.discount)}
          onChange={(e) => onDiscoutChange(e.value)}
        />
      </div>

      <div className="p-field p-col">
        <label htmlFor="amountPayable"> Amount payable </label>
        <InputNumber
          id="amountPayable"
          value={defaultNum(billDetails.amountPayable)}
        />
      </div>

      <div className="p-field p-col">
        <label htmlFor="paid"> Amount paid </label>
        <InputNumber
          id="paid"
          value={defaultNum(billDetails.paid)}
          onChange={(e) => onAmountPaidChange(e.value)}
        />
      </div>

      <div className="p-field p-col">
        <label htmlFor="due"> Due </label>
        <InputNumber id="due" value={defaultNum(billDetails.due)} />
      </div>
    </div>
  );
}
