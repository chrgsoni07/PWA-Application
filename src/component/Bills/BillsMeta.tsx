import { Calendar } from "primereact/calendar";
import { InputNumber } from "primereact/inputnumber";
import { InputText } from "primereact/inputtext";
import React, { useState } from "react";
export function BillsMeta({ invoiceDate, setInvoiceDate }: any) {
  const [previousAmount, setPreviousAmount] = useState();
  const [advanceAmount, setAdvanceAmount] = useState();
  return (
    <div className="p-col-4">
      <div className="p-formgrid p-grid">
        <div className="p-field p-col">
          <label htmlFor="invoiceNo">Invoice no</label>
          <InputText value={101} />
        </div>

        <div className="p-field p-col">
          <label htmlFor="invoiceDate">Invoice Date</label>
          <Calendar
            id="invoiceDate"
            value={invoiceDate}
            onChange={(e) => setInvoiceDate(e.value)}
            showIcon
          />
        </div>
      </div>

      <div className="p-formgrid  p-grid">
        <div className="p-field p-col">
          <label htmlFor="advanceAmount">Advance Amount</label>
          <InputNumber
            id="advanceAmount"
            value={advanceAmount}
            onChange={(e) => setAdvanceAmount(e.value)}
          />
        </div>

        <div className="p-field p-col">
          <label htmlFor="previousAmount">Previous Amount</label>
          <InputNumber
            id="previousAmount"
            value={previousAmount}
            onChange={(e) => setPreviousAmount(e.value)}
          />
        </div>
      </div>
    </div>
  );
}
