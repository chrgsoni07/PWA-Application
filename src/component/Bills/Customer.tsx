import { Dropdown } from "primereact/dropdown";

function Customer({ selectedCustomer, customers, setSelectedCustomer }: any) {
  return (
    <div className="p-col-8">
      {/* drop down */}
      <div className="p-formgrid p-grid">
        <div className="p-field p-col-12">
          <label htmlFor="customerSelect">Select customer</label>
          <Dropdown
            id="customerSelect"
            value={selectedCustomer}
            options={customers}
            onChange={(e) => setSelectedCustomer(e.value || {})}
            optionLabel="name"
            filter
            showClear
            filterBy="name"
            placeholder="Select a customer"
            style={{
              width: "100%",
            }}
          />
        </div>
      </div>

      <div className="p-fluid p-formgrid p-grid">
        <div className="p-field p-col">
          <label htmlFor="name">Name: </label>
          <span id="name">{selectedCustomer?.name}</span>
        </div>

        <div className="p-field p-col">
          <label htmlFor="mobile">Mobile: </label>
          <span id="mobile">{selectedCustomer?.mobile}</span>
        </div>

        <div className="p-field p-col">
          <label htmlFor="place">Place: </label>
          <span id="place">{selectedCustomer?.place}</span>
        </div>

        <div className="p-field p-col">
          <label htmlFor="address">Address: </label>
          <span id="address">{selectedCustomer?.address}</span>
        </div>
      </div>
    </div>
  );
}
export default Customer;
