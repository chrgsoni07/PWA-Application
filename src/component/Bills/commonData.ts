export const itemType = [
  { label: "Gold", value: "gold" },
  { label: "Silver", value: "silver" },
  { label: "Fixed", value: "fixed" },
];
