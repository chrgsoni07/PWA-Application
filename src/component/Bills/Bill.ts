import { CustomerType } from "component/Customers/types";
import { BillDetails } from "./BillDetails";
import { NewItem } from "./NetItem";
import { OldItem } from "./OldItem";

export type Bill = {
  billNo: number;
  invoiceDate: string;
  customer: CustomerType;
  newItems: NewItem[];
  oldItems: OldItem[];
  billDetail: BillDetails;
};
