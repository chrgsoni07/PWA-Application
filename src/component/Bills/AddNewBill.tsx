import { CustomerType } from "component/Customers/types";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import { TabPanel, TabView } from "primereact/tabview";
import { FC, useEffect, useState } from "react";
import { sum } from "utils/number.utils";
import { BillsMeta } from "./BillsMeta";
import { BillTotals } from "./BillTotals";
import { defaultBill } from "./commonData";
import Customer from "./Customer";
import { NewItems } from "./NewItems";
import { OldItems } from "./OldItems";
import { Bill, BillDetails, NewItem, OldItem } from "./types";

type AddNewBillProps = {
  displayDialog: boolean;
  bill: Bill;
  setDisplayDialog: (flag: boolean) => void;
  setBill: (bill: Bill) => void;
};
export const AddNewBill: FC<AddNewBillProps> = ({
  displayDialog,
  setDisplayDialog,
  bill,
  setBill,
}) => {
  const [newItems, setNewItems] = useState<NewItem[]>([]);
  const [oldItems, setOldItems] = useState<OldItem[]>([]);
  const [invoiceDate, setInvoiceDate] = useState<Date>(new Date());

  const [selectedCustomer, setSelectedCustomer] = useState<CustomerType>();
  const [billDetails, setBillDetails] = useState<BillDetails>({
    newTotal: 0,
    oldTotal: 0,
    oldNewDifference: 0,
    discount: 0,
    paid: 0,
    due: 0,
    amountPayable: 0,
  });

  const onDiscoutChange = (discount: number) => {
    let amountPayable = billDetails.oldNewDifference - discount;

    if (discount) {
      setBillDetails({ ...billDetails, discount, amountPayable });
    }
  };

  const onAmountPaidChange = (paid: number) => {
    let due = billDetails.amountPayable - paid;
    setBillDetails({ ...billDetails, paid, due });
  };

  useEffect(() => {
    setNewItems(bill.newItems);
    setOldItems(bill.oldItems);
    bill.billDetail && setBillDetails(bill.billDetail);
    setInvoiceDate(bill.invoiceDate);
  }, [bill]);

  useEffect(() => {
    const newTotal = sum(newItems);
    const oldTotal = sum(oldItems);
    const oldNewDifference = newTotal - oldTotal;

    setBillDetails({
      ...billDetails,
      oldTotal,
      newTotal,
      oldNewDifference,
      amountPayable: oldNewDifference > 0 ? oldNewDifference : 0,
    });
  }, [newItems, oldItems]);

  const onHide = () => {
    setDisplayDialog(false);
    setBill(defaultBill());
  };

  const onRowEditInit = () => {};

  const onRowEditCancel = () => {};

  const onSave = () => {
    const newBill = {
      ...bill,
      invoiceDate,
      customer: selectedCustomer,
      newItems,
      oldItems,
      billDetail: billDetails,
    };
    console.log("newItems", newItems);
    console.log("updated bill", newBill);
    setBill(newBill);
    console.log(newBill); //  saveBillToFirestore(newBill);
  };

  const footer = (
    <span
      style={{
        display: "table",
        margin: "0 auto",
      }}
    >
      <Button
        label="Save"
        icon="pi pi-save"
        style={{
          marginRight: ".25em",
        }}
        onClick={onSave}
      />
      <Button label="Draft" icon="pi pi-check" className="p-button-secondary" />
    </span>
  );
  return (
    <Dialog
      visible={displayDialog}
      onHide={onHide}
      header="New Bill"
      footer={footer}
      modal
      breakpoints={{
        "960px": "75vw",
        "640px": "100vw",
      }}
      style={{
        width: "50vw",
      }}
      maximized={true}
    >
      <div className="p-grid">
        <Customer
          selectedCustomer={selectedCustomer}
          setSelectedCustomer={setSelectedCustomer}
        />

        <BillsMeta invoiceDate={invoiceDate} setInvoiceDate={setInvoiceDate} />
      </div>

      <TabView>
        <TabPanel header="New Item">
          <NewItems
            newItems={newItems}
            onRowEditInit={onRowEditInit}
            onRowEditCancel={onRowEditCancel}
            billDetails={billDetails}
            setNewItems={setNewItems}
          />
        </TabPanel>
        <TabPanel header="Old Item">
          <OldItems
            setOldItems={setOldItems}
            oldItems={oldItems}
            onRowEditInit={onRowEditInit}
            onRowEditCancel={onRowEditCancel}
            billDetails={billDetails}
          />
        </TabPanel>
      </TabView>
      <BillTotals
        onDiscoutChange={onDiscoutChange}
        onAmountPaidChange={onAmountPaidChange}
        billDetails={billDetails}
      />
    </Dialog>
  );
};
