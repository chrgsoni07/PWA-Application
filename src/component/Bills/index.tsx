import { AddNewBill } from "./AddBill/AddNewBill";
import ViewBill from "./ViewBill";
import { Button } from "primereact/button";
import { Card } from "primereact/card";
import { TabView, TabPanel } from "primereact/tabview";
import React, { useState, useEffect } from "react";
import "./DataTableDemo.css";
import { Bill } from "./types";
import { Column } from "primereact/column";
import { DataTable } from "primereact/datatable";
import { createNextState } from "@reduxjs/toolkit";
import { db, getBills } from "api";
import { Dialog } from "primereact/dialog";
import { useToast } from "toasts";

const Bills = () => {
  const [activeIndex, setActiveIndex] = useState(0);
  const [displayDialog, setDisplayDialog] = useState(false);
  const [displayViewDialog, setDisplayViewDialog] = useState(false);
  const [savedBills, setSavedBills] = useState<Bill[]>([]);
  const [draftBills, setDraftBills] = useState<Bill[]>([]);
  const [bill, setBill] = useState<Bill>({} as Bill);
  const { toastSuccess, toastError } = useToast();

  useEffect(() => {
    getBills().then((bills) => setSavedBills(bills));
    let draftBills = JSON.parse(localStorage.getItem("draftBills") || "[]");
    console.log("draftbills => ", draftBills);
    setDraftBills(draftBills);
  }, []);

  const header = () => {
    <Button label="New" icon="pi pi-plus" className="p-button-sm" />;
  };

  const displayModel = () => {
    setDisplayDialog(true);
  };

  const dateBodyTemplate = (rowData: any) => {
    var invoiceDate = rowData.invoiceDate;
    return invoiceDate.toLocaleDateString("en-In");
  };

  const actionBodyTemplate = (rowData: any) => {
    function viewBill(rowData: any): void {
      setBill({ ...rowData, invoiceDate: dateBodyTemplate(rowData) });
      setDisplayViewDialog(true);
    }

    function editBill(rowData: any): void {
      setBill({ ...rowData, invoiceDate: rowData.invoiceDate });
      setDisplayDialog(true);
    }

    function confirmDeleteBill(rowData: any): void {
      alert(JSON.stringify(rowData));
      db.collection("bills")
        .doc(rowData.id)
        .delete()
        .then(() => {
          toastSuccess("bill successfully deleted");
          setSavedBills(
            createNextState(savedBills, (draft) =>
              draft.filter((i) => i.id !== rowData.id)
            )
          );
        })
        .catch((error: Error) => {
          toastError("Error deleting bill" + error.message);
        });
    }

    return (
      <>
        <Button
          icon="pi pi-eye"
          className="p-button-rounded p-button-help p-mr-2"
          onClick={() => viewBill(rowData)}
        />
        <Button
          aria-label="editBill"
          icon="pi pi-pencil"
          className="p-button-rounded p-button-success p-mr-2"
          onClick={() => editBill(rowData)}
        />
        <Button
          icon="pi pi-trash"
          className="p-button-rounded p-button-warning"
          onClick={() => confirmDeleteBill(rowData)}
        />
      </>
    );
  };

  return (
    <>
      <Card header={header}>
        <Button
          label="New"
          icon="pi pi-plus"
          className="p-button-success p-mr-2"
          onClick={() => displayModel()}
        />
        <TabView
          activeIndex={activeIndex}
          onTabChange={(e) => setActiveIndex(e.index)}
        >
          <TabPanel header="Previous Bills">
            <div className="card">
              <DataTable
                value={savedBills}
                paginator
                rows={10}
                rowsPerPageOptions={[5, 10, 25]}
                paginatorTemplate="FirstPageLink PrevPageLink PageLinks NextPageLink LastPageLink CurrentPageReport RowsPerPageDropdown"
                currentPageReportTemplate="Showing {first} to {last} of {totalRecords} items"
                selectionMode="single"
                dataKey="id"
                className="p-datatable-gridlines p-datatable-sm"
              >
                <Column
                  field="id"
                  header="Id"
                  sortable
                  body={(_: any, prop: any) => prop.rowIndex + 1}
                ></Column>
                <Column
                  field="billNo"
                  header="Bill No"
                  data-testid="test"
                  key="test3"
                  filter
                  sortable
                  filterPlaceholder="Search by bill no"
                ></Column>
                <Column
                  field="invoiceDate"
                  header="Date"
                  body={dateBodyTemplate}
                  filter
                  filterPlaceholder="Search by date"
                ></Column>
                <Column
                  field="customer.name"
                  header="Customer"
                  filter
                  sortable
                  filterPlaceholder="Search by customer no"
                ></Column>
                <Column
                  field="billDetail.amountPayable"
                  header="Amount Payable"
                ></Column>
                <Column field="billDetail.paid" header="Paid"></Column>
                <Column field="billDetail.due" header="Due"></Column>
                <Column body={actionBodyTemplate}></Column>
              </DataTable>
            </div>
          </TabPanel>
          <TabPanel header="Draft">
            <div className="card">
              <DataTable
                value={draftBills}
                selectionMode="single"
                dataKey="id"
                className="p-datatable-gridlines p-datatable-sm"
              >
                <Column
                  field="id"
                  header="Id"
                  sortable
                  body={(_: any, prop: any) => prop.rowIndex + 1}
                ></Column>
                <Column
                  field="invoiceDate"
                  header="Date"
                  //    body={dateBodyTemplate}
                ></Column>
                <Column field="customer.name" header="Customer"></Column>
                <Column
                  field="billDetail.amountPayable"
                  header="Amount Payable"
                ></Column>
                <Column field="billDetail.paid" header="Paid"></Column>
                <Column field="billDetail.due" header="Due"></Column>
                <Column body={actionBodyTemplate}></Column>
              </DataTable>
            </div>
          </TabPanel>
        </TabView>
      </Card>

      <AddNewBill
        bill={bill}
        setBill={setBill}
        displayDialog={displayDialog}
        setDisplayDialog={setDisplayDialog}
      />
      <Dialog
        visible={displayViewDialog}
        onHide={() => setDisplayViewDialog(false)}
        header="R K JEWELLERS JAWAD"
        modal
        breakpoints={{
          "960px": "75vw",
          "640px": "100vw",
        }}
        style={{
          width: "50vw",
        }}
        maximized={true}
      >
        <ViewBill bill={bill} />
      </Dialog>
    </>
  );
};

export default Bills;
