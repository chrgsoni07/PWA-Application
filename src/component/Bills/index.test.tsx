import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ToastsProvider } from "toasts";
import Bills from ".";

jest.mock("api");

describe("Bills", () => {
  it("should render", async () => {
    const { asFragment } = render(
      <ToastsProvider>
        <Bills />
      </ToastsProvider>
    );
    const firstRender = asFragment();
    await screen.findByRole("button", { name: "editBill" });
    userEvent.click(screen.getByRole("button", { name: "New" }));
    userEvent.click(screen.getByRole("button", { name: /close/i }));

    const tableloaded = asFragment();
    expect(firstRender).toMatchDiffSnapshot(tableloaded);
    userEvent.click(screen.getByRole("button", { name: "editBill" }));

    expect(screen.getByText("2020202020")).toBeInTheDocument();
    const newItemTab = screen.getByRole("tabpanel", { name: "New Item" });
    expect(within(newItemTab).getAllByRole("row")).toHaveLength(4);
    // const dialogOpened = asFragment();
    // expect(tableloaded).toMatchDiffSnapshot(dialogOpened);
  });
});
