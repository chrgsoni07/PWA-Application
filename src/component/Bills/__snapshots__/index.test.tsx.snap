// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bills should render 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -254,11 +254,102 @@
                            />
                          </tr>
                        </thead>
                        <tbody
                          class=\\"p-datatable-tbody\\"
+                       >
+                         <tr
+                           class=\\"p-selectable-row\\"
+                           role=\\"row\\"
+                           tabindex=\\"0\\"
+                         >
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             1
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             1
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             30/5/2021
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             Chaupat Raja
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             35200
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             30000
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             5200
+                           </td>
+                           <td
+                             class=\\"\\"
+                             role=\\"cell\\"
+                           >
+                             <button
+                               class=\\"p-button p-component p-button-rounded p-button-help p-mr-2 p-button-icon-only\\"
+                             >
+                               <span
+                                 class=\\"p-button-icon p-c pi pi-eye\\"
+                               />
+                               <span
+                                 class=\\"p-button-label p-c\\"
+                               >
+                                  
+                               </span>
+                             </button>
+                             <button
+                               aria-label=\\"editBill\\"
+                               class=\\"p-button p-component p-button-rounded p-button-success p-mr-2 p-button-icon-only\\"
+                             >
+                               <span
+                                 class=\\"p-button-icon p-c pi pi-pencil\\"
+                               />
+                               <span
+                                 class=\\"p-button-label p-c\\"
+                               >
+                                  
+                               </span>
+                             </button>
+                             <button
+                               class=\\"p-button p-component p-button-rounded p-button-warning p-button-icon-only\\"
+                             >
+                               <span
+                                 class=\\"p-button-icon p-c pi pi-trash\\"
                                />
+                               <span
+                                 class=\\"p-button-label p-c\\"
+                               >
+                                  
+                               </span>
+                             </button>
+                           </td>
+                         </tr>
+                       </tbody>
                      </table>
                    </div>
                    <div
                      class=\\"p-paginator p-component p-paginator-bottom\\"
                    >
@@ -309,11 +400,11 @@
                        />
                      </button>
                      <span
                        class=\\"p-paginator-current\\"
                      >
-                       Showing 0 to 0 of 0 items
+                       Showing 1 to 1 of 1 items
                      </span>
                      <div
                        class=\\"p-dropdown p-component p-inputwrapper p-inputwrapper-filled\\"
                      >
                        <div"
`;
